driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -378,6 +382,17 @@ private boolean addConnectionIfUnderMaximum() {
         }
     }
 
+    private PooledConnection tryResurrectFromTrash() {
+        long now = System.currentTimeMillis();
+        for (PooledConnection connection : trash)
+            if (connection.maxIdleTime >= now && connection.state.compareAndSet(TRASHED, OPEN)) {
```

Comment:
If this is the reason why the connection was initially trashed, then it went through `replaceConnection`, which sets `maxIdleTime` in the past. I thought this was simpler than adding a 4th state (something like `DEFINITIVELY_TRASHED`), but if this is too obscure I can reconsider.

But there is another scenario: the connection was trashed because of a low inFlight, and then some of the remaining queries were cancelled or timed out, which caused `maxAvailableStreams` to hit the limit while in the trash. This is probably very rare but it can happen, so I'll add the check, thanks.